### PR TITLE
WIP: TDFB Improve tdbf directivity index and WNG

### PR DESCRIPTION
**Pull request by ShriramShastry**
Added singular value decomposition and pwelch algorithms
	for diagonal matrix calculation. Added filters for side
	lobe artifacts.

Signed-off-by: ShriramShastry <malladi.sastry@intel.com>

### CODE REVIEW

File: tools/tune/tdfb/bf_design.m [hunk 1]
```diff
@@ -119,6 +123,37 @@
...
+                %Power Spectral Density estimate via Welch's method
+                if exist('OCTAVE_VERSION', 'builtin')
+                    [Pxyz,freq] = pwelch(sinc(2*pi*f*lnm/bf.c),[], [], N, fs, 0);
+                    Gamma_vv(:,n,m) = pwelch(Pxyz  ,[], [], N, fs, 0);
+                else
+                    [Pxyz,freq] = pwelch(2*pi*f*lnm/bf.c,hamming(510), [], N, fs);
+                    Gamma_vv(:,n,m) = pwelch(Pxyz  ,hamming(510), [], N, fs);
+                end
+            elseif isequal(DEBUG_DIAG_LOAD_ON,3)  % V is beamformer matrix
+                %Power Spectral Density estimate via Welch's method -centered two-sided PSD
```
Comment (cujomalainey) [inline tools/tune/tdfb/bf_design.m:139]: this code block is duplicated below, please either refactor or turn into a function

Comment (ShriramShastry) [reply tools/tune/tdfb/bf_design.m]: Thanks! I have removed  the code section since I didn't observe improvement in directive index and with white noise gain.

File: tools/tune/tdfb/bf_design.m [hunk 2]
```diff
@@ -20,11 +20,11 @@
 	error('The steer_az and steer_el lengths need to be equal.');
 end
 
-if length(find(bf.steer_az > 180)) || length(find(bf.steer_az < -180))
+if ~isempty(find(bf.steer_az > 180, 1)) || ~isempty(find(bf.steer_az < -180, 1))
 	error('The steer_az angles need to be -180 to +180 degrees');
 end
 
-if length(find(bf.steer_el > 90)) || length(find(bf.steer_el < -90))
+if ~isempty(find(bf.steer_el > 90, 1)) || ~isempty(find(bf.steer_el < -90, 1))
```
Comment (singalsu) [inline tools/tune/tdfb/bf_design.m:27]: What is the issue that ~isempty(find(),1) fixes? It's more complicated then length(find()). Also, the generic script fixes not related to design improvement should be in another commit.

File: tools/tune/tdfb/bf_design.m [hunk 3]
```diff
@@ -189,6 +197,18 @@
...
+elseif isequal(DEBUG_SELT_FILT_ON,2)
+    win = hamming(bf.fir_length);
+elseif isequal(DEBUG_SELT_FILT_ON,3)
+    sidelobe = -30;
+    nbar = 4;
+    win = taylorwin(bf.fir_length, nbar, sidelobe);
+elseif isequal(DEBUG_SELT_FILT_ON,4)
+    sidelobe = 30;
+    win = chebwin(bf.fir_length, sidelobe);
+end
```
Comment (singalsu) [inline tools/tune/tdfb/bf_design.m:211]: The above could be a switch lower(bf_window), case 'kaiser', case 'hann', etc. The parameters for them could be like bf_kaiser_beta, bf_taylorwin_sidelobe, bf_taylorwin_nbar = 4, etc. You could set the bf_window to 'hann' in bf_defaults if that gave the best results.

File: tools/tune/tdfb/bf_design.m [hunk 4]
```diff
@@ -119,6 +123,10 @@
 			+(bf.mic_y(n) - bf.mic_y(m))^2 ...
 			+(bf.mic_z(n) - bf.mic_z(m))^2);
             Gamma_vv(:,n,m) = sinc(2*pi*f*lnm/bf.c);
+		if isequal(DEBUG_DIAG_LOAD_ON,1)
```
Comment (singalsu) [inline tools/tune/tdfb/bf_design.m:126]: if DEBUG_DIAG_LOAD_ON == 1 would work too. Though I think the proposed diagonal load could be made default without option. Or make it an bf_defaults.m set option if you think enabled/disabled would make sense.

File: tools/tune/tdfb/bf_design.m [hunk 5]
```diff
@@ -246,6 +266,9 @@
     denom2 = denom1 * W_w;
     di = num / denom2;
     bf.di_db(iw) = 10*log10(abs(di));
+    if isequal(DEBUG_DIAG_LOAD_ON,1)
+        bf.di_db(iw) = log(abs(di)*fs/N)*1e+1/3;
```
Comment (singalsu) [inline tools/tune/tdfb/bf_design.m:270]: I still wonder why the di_db calculation needs to change here. Also 10*log10() is nicer to read than log10()*1e+1. What is the divide by 3 for?

File: tools/tune/tdfb/bf_design.m [hunk 6]
```diff
@@ -505,7 +531,7 @@ function ch_legend(n)
 
 function myaudiowrite(fn, x, fs)
 [~, ~, ext] = fileparts(fn);
-if strcmp(lower(ext), '.raw')
+if strcmpi(ext, '.raw')
```
Comment (singalsu) [inline tools/tune/tdfb/bf_design.m:534]: Nice, but this fits the overall script improvement commit.

File: tools/tune/tdfb/bf_design.m (hunk 3 above)
Comment (ShriramShastry) [reply tools/tune/tdfb/bf_design.m:212]: Sure !

File: tools/tune/tdfb/bf_design.m (hunk 5 above)
Comment (ShriramShastry) [reply tools/tune/tdfb/bf_design.m]: Sure - Your proposal make sense. Thanks

File: tools/tune/tdfb/bf_design.m (hunk 4 above)
Comment (ShriramShastry) [reply tools/tune/tdfb/bf_design.m]: Sure, I will make it as default

**Comment by cujomalainey** [comment]
From a code standpoint I can give :+1: but I cannot verify math

File: tools/tune/tdfb/bf_design.m [hunk 7]
```diff
@@ -17,15 +17,15 @@
 mkdir_check('data');
 
 if length(bf.steer_az) ~= length(bf.steer_el)
-	error('The steer_az and steer_el lengths need to be equal.');
+    error('The steer_az and steer_el lengths need to be equal.');
```
Comment (singalsu) [inline tools/tune/tdfb/bf_design.m:20]: I think you should configure the editor for indent by 8 with tab. We don't have a rule for .m but that would be same as for C code.

File: tools/tune/tdfb/bf_design.m (hunk 2 above)
Comment (singalsu) [reply tools/tune/tdfb/bf_design.m:27]: Maybe this would be nicest way to error with illegal angles `if max(bf.steer_az) > 180 || min(bf.steer_az) < -180`

File: tools/tune/tdfb/bf_design.m [hunk 8]
```diff
@@ -200,7 +199,21 @@
...
 start = center_idx - floor(bf.fir_length/2);
-win0 = kaiser(bf.fir_length,bf.fir_beta);
+switch lower(bf.type_filt)
+    case 'hann'
+        win0 = hann (bf.fir_length);
+    case 'hamming'
+        win0 = hamming(bf.fir_length);
+    case 'taylorwin'
+        sidelobe = -30;
+        nbar = 4;
```
Comment (singalsu) [inline tools/tune/tdfb/bf_design.m:209]: these to bf_defaults as bf.taylorwin_sidelobe, bf.taylorwin_nbar.

File: tools/tune/tdfb/bf_design.m [hunk 9]
```diff
@@ -200,7 +199,21 @@
...
+        win0 = taylorwin(bf.fir_length, nbar, sidelobe);
+    case 'chebwin'
+        sidelobe = 30;
+        win0 = chebwin(bf.fir_length, sidelobe);
+    case 'kaiser'
+        win0 = kaiser(bf.fir_length,bf.fir_beta);
```
Comment (singalsu) [inline tools/tune/tdfb/bf_design.m:215]: better to rename bf.fir_beta to bf.kaiser_beta

File: tools/tune/tdfb/example_line_array.m [hunk 10]
```diff
@@ -43,6 +43,7 @@ function example_line_array()
 bf.output_stream_mix    = [0 0]; % Mix both filters to stream 0
 bf.num_output_channels = 2;      % Two channels
 bf.num_output_streams = 1;       % One sink stream
+bf.type_filt = 'hann';           % filter
```
Comment (singalsu) [inline tools/tune/tdfb/example_line_array.m:46]: you can put this to bf_defaults to avoid to do it every time

**Comment by ShriramShastry** [comment]
I have an idea why `Zephyr builds` didn't go through

**Review by lgirdwood — APPROVED** [review]
code style looks fine, but @singalsu we need you approval to merge.

**Review by singalsu — APPROVED** [review]
It looks OK but remove WIP from patch title. Thanks!

**Comment by lgirdwood** [comment]
SOFCI TEST

**Comment by lgirdwood** [comment]
Rerunning CI.

**Comment by ShriramShastry** [comment]
@lgirdwood ,  why  `zephyr-build` still failing ?

**Comment by lgirdwood** [comment]
> @lgirdwood , why `zephyr-build` still failing ?

There are some recent updates that fix this, can you update to latest main and rebase then force push. This should solve all the CI issues.

**Comment by aiChaoSONG** [comment]
@ShriramShastry https://github.com/thesofproject/sof/commit/a12145431d3c3b780e201c00b91882d412ceda71 is required to make zephyr build pass. make sure it is in your base

**Comment by ShriramShastry** [comment]
> a121454

Thanks @aiChaoSONG @lgirdwood .

**Comment by lgirdwood** [comment]
CI showing an unrelated keyboard error for TGL suspend TIMEOUT, but otherise this test looks fine.  https://sof-ci.01.org/sofpr/PR4581/build10138/devicetest/?model=TGLH_RVP_HDA&testcase=check-suspend-resume-with-playback
